### ISSUE TITLE
Wrong casting (mixed precision) in attention layers

### ISSUE THREAD

**Issue by shkarupa-alex** [tensorflow/tensorflow#46064]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Google Colab
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): Google Colab
- TensorFlow version (use command below): v2.4.0-0-g582c8d236cb 2.4.0
- Python version: 3
- Bazel version (if compiling from source): No
- GCC/Compiler version (if compiling from source): No
- CUDA/cuDNN version: No
- GPU model and memory: No

**Describe the current behavior**
When using Attention or AdditiveAttention with mixed precision policy issue occurred due to wrong casting (mask casted to floatx but should be casted to scores.dtype)

**Describe the expected behavior**
Layers should work without issues with mixed_fp16

**Standalone code to reproduce the issue**
https://colab.research.google.com/drive/1R2MKXAIrmYBBGkij11m9aG7hLm2AEHjF?usp=sharing

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.
The issue comes from here https://github.com/tensorflow/tensorflow/blob/v2.4.0/tensorflow/python/keras/layers/dense_attention.py#L129
Should be this:
```python
scores -= 1.e9 * math_ops.cast(padding_mask, dtype=scores.dtype)
```

**Comment by ravikyram** [comment]
I have tried in colab with TF version 2.4, Nightly version(`2.5.0-dev20201229`) and was able to reproduce the issue. Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/7b6b6b54fbf4b7df517cb537b726e371/untitled588.ipynb). Thanks!

**Comment by yongtang** [comment]
Added a PR #46321 for the fix.
